aesd-char-driver/main.c

```diff
@@ -179,12 +188,149 @@ ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
 		mutex_unlock(&dev->lock);
 		return retval;
 }
+
+loff_t aesd_llseek (struct file * filp, loff_t offset, int whence){
+	
+	
+	//Get the aesd device structure
+	struct aesd_dev *dev = filp->private_data;
+	loff_t new_position;
+	size_t total_buffer_size = 0;
+	uint8_t index;
+	struct aesd_buffer_entry *entryptr;
+	ssize_t retval = 0;
+
+
+	//First we lock the structure
+	int res = mutex_lock_interruptible(&(dev->lock));
+	if(res){
+		return -ERESTARTSYS;
+	}
+	
+	//Calculate the size of the circular buffer by iterating through the entries
+	AESD_CIRCULAR_BUFFER_FOREACH(entryptr,&dev->circular_buffer,index){
+		if(entryptr->buffptr){
+			total_buffer_size+= entryptr->size;
+		}	
+	}
+	//Unlock the mutex
+	mutex_unlock(&dev->lock);
+	PDEBUG("Total size of buffer is %zu",total_buffer_size);
+	//Now call generic llseek
+	new_position = fixed_size_llseek(filp, offset, whence,total_buffer_size);
+	
+	PDEBUG("New position is %zu",new_position);
+	if(new_position < 0){
+		retval = -EINVAL;
+	}
+	
+	
+	return retval;
+}
+
+long aesd_ioctl(struct file *filp, unsigned int cmd, unsigned long arg){
+	
+	 //Get the aesd device structure
+	struct aesd_dev *dev = filp->private_data;
+	//Cast the incoming argument
+	struct aesd_seekto temp_aesd_seekto ;
+	uint8_t cmd_ptr=0;
+	int cmds_scanned = -1;
+	ssize_t retval = 0;
+	long new_fpos = 0;
+	
+	PDEBUG("AESD IOCTL successfully called");
+	
+	//Check to make sure the magic number used is matched
+	if(_IOC_TYPE(cmd) != AESD_IOC_MAGIC){
+		return -ENOTTY;
+	}
+	PDEBUG("Magic number checked passed");
+	//Check to make sure the number of commands is less than whats requested
+	 if (_IOC_NR(cmd) > AESDCHAR_IOC_MAXNR){
+		 return -ENOTTY; 
+	 } 
+	 PDEBUG("Num commands check passed passed");
+	 //Check to make sure it is a Read and Write direction
+	 if(_IOC_DIR(cmd) != (_IOC_READ |_IOC_WRITE)){
+		 return -ENOTTY;
+	 }
+	  PDEBUG("Read and write direction check passed");
```

Comment:
@Silex93 should have one additional check here to ensure `cmd` is actually `AESDCHAR_IOCSEEKTO`